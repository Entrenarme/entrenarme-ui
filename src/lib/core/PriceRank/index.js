// @flow
import * as React from 'react';
import Slider, { Range } from 'rc-slider';
import times from 'lodash/times';

import ExtendedRange from './styles';

type Props = {
  currency?: string,
  priceRank?: Array<number>,
  callback: Function,
  slider: boolean,
  min: number,
  max: number,
  defaultValue: number,
  step: number,
  markTooltip: string,
};

type State = {};

class PriceRank extends React.Component<Props, State> {
  static defaultProps = {
    currency: 'â‚¬',
    priceRank: [0, 3],
    slider: false,
    min: 0,
    max: 3,
    defaultValue: 0,
    step: 1,
    markTooltip: '',
  };

  state = {
    Cmp: this.props.slider ? ExtendedRange(Slider) : ExtendedRange(Range),
  };

  renderCurrencyMark = (repetitions: number) => {
    const { currency } = this.props;

    const iterableArray = times(repetitions, String);

    return iterableArray.map(element => currency).join('');
  };

  handleChange = (rank: Array<number>) => {
    this.props.callback(rank);
  };

  renderMarks = () => {
    const marks = {};

    const iterableArray = times(4, String);

    iterableArray.forEach(
      element =>
        (marks[element] = this.renderCurrencyMark(parseInt(element, 10) + 1)),
    );

    return marks;
  };

  render() {
    const {
      priceRank,
      slider,
      min,
      max,
      defaultValue,
      step,
      markTooltip,
      value,
    } = this.props;
    const { Cmp } = this.state;
    if (slider) {
      if (value !== null || typeof value !== 'undefined') {
        return (
          <Cmp
            min={min}
            max={max}
            step={step}
            defaultValue={defaultValue}
            value={value}
            onChange={this.handleChange}
            marks={{
              [min]: `${min}${markTooltip}`,
              [max]: `${max}${markTooltip}`,
            }}
          />
        );
      }
      return (
        <Cmp
          min={min}
          max={max}
          step={step}
          defaultValue={defaultValue}
          onChange={this.handleChange}
          marks={{
            [min]: `${min}${markTooltip}`,
            [max]: `${max}${markTooltip}`,
          }}
        />
      );
    }
    return (
      <Cmp
        defaultValue={priceRank}
        min={min}
        max={max}
        value={value}
        onChange={this.handleChange}
        marks={this.renderMarks()}
      />
    );
  }
}

export default PriceRank;
