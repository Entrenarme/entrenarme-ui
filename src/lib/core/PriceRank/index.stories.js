// @flow
import * as React from 'react';
import { storiesOf } from '@storybook/react';

import PriceRank from './index';

storiesOf('PriceRank', module)
  .add('default', () => <PriceRank callback={rank => console.log(rank)} />)
  .add('with dollar currency', () => (
    <PriceRank currency="$" callback={rank => console.log(rank)} />
  ))
  .add('with default price rank', () => (
    <PriceRank
      currency="â‚¬"
      priceRank={[1, 2]}
      callback={rank => console.log(rank)}
    />
  ))
  .add('Slider', () => (
    <PriceRank
      slider
      min={0.5}
      max={30}
      step={0.5}
      markTooltip="km"
      callback={rank => console.log(rank)}
    />
  ));
