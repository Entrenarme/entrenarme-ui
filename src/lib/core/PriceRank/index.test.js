import * as React from 'react';
import { render } from 'react-testing-library';
import 'react-testing-library/cleanup-after-each';

import PriceRank from './index';

test('renders price rank', () => {
  const { container } = render(<PriceRank currency="$" callback={() => {}} />);

  expect(container.firstChild).toMatchSnapshot();
});
