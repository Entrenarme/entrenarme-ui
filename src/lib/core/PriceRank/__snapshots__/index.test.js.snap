// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders price rank 1`] = `
<div
  class="rc-slider rc-slider-with-marks sc-bdVaJa dvPGHo"
>
  <div
    class="rc-slider-rail"
  />
  <div
    class="rc-slider-track rc-slider-track-1"
    style="left: 0%; width: 100%;"
  />
  <div
    class="rc-slider-step"
  >
    <span
      class="rc-slider-dot rc-slider-dot-active"
      style="left: 0%;"
    />
    <span
      class="rc-slider-dot rc-slider-dot-active"
      style="left: 33.33333333333333%;"
    />
    <span
      class="rc-slider-dot rc-slider-dot-active"
      style="left: 66.66666666666666%;"
    />
    <span
      class="rc-slider-dot rc-slider-dot-active"
      style="left: 100%;"
    />
  </div>
  <div
    role="slider"
    tabindex="0"
    aria-valuemin="0"
    aria-valuemax="3"
    aria-valuenow="0"
    aria-disabled="false"
    class="rc-slider-handle rc-slider-handle-1"
    style="left: 0%;"
  />
  <div
    role="slider"
    tabindex="0"
    aria-valuemin="0"
    aria-valuemax="3"
    aria-valuenow="3"
    aria-disabled="false"
    class="rc-slider-handle rc-slider-handle-2"
    style="left: 100%;"
  />
  <div
    class="rc-slider-mark"
  >
    <span
      class="rc-slider-mark-text rc-slider-mark-text-active"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 0%;"
    >
      $
    </span>
    <span
      class="rc-slider-mark-text rc-slider-mark-text-active"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 33.33333333333333%;"
    >
      $$
    </span>
    <span
      class="rc-slider-mark-text rc-slider-mark-text-active"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 66.66666666666666%;"
    >
      $$$
    </span>
    <span
      class="rc-slider-mark-text rc-slider-mark-text-active"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 100%;"
    >
      $$$$
    </span>
  </div>
</div>
`;

exports[`renders price rank 2`] = `
<div
  class="rc-slider rc-slider-with-marks sc-bdVaJa dvPGHo"
>
  <div
    class="rc-slider-rail"
  />
  <div
    class="rc-slider-track rc-slider-track-1"
    style="left: 66.66666666666666%; width: 33.33333333333334%;"
  />
  <div
    class="rc-slider-step"
  >
    <span
      class="rc-slider-dot"
      style="left: 0%;"
    />
    <span
      class="rc-slider-dot"
      style="left: 33.33333333333333%;"
    />
    <span
      class="rc-slider-dot rc-slider-dot-active"
      style="left: 66.66666666666666%;"
    />
    <span
      class="rc-slider-dot rc-slider-dot-active"
      style="left: 100%;"
    />
  </div>
  <div
    role="slider"
    tabindex="0"
    aria-valuemin="0"
    aria-valuemax="3"
    aria-valuenow="2"
    aria-disabled="false"
    class="rc-slider-handle rc-slider-handle-1"
    style="left: 66.66666666666666%;"
  />
  <div
    role="slider"
    tabindex="0"
    aria-valuemin="0"
    aria-valuemax="3"
    aria-valuenow="3"
    aria-disabled="false"
    class="rc-slider-handle rc-slider-handle-2"
    style="left: 100%;"
  />
  <div
    class="rc-slider-mark"
  >
    <span
      class="rc-slider-mark-text"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 0%;"
    >
      €
    </span>
    <span
      class="rc-slider-mark-text"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 33.33333333333333%;"
    >
      €€
    </span>
    <span
      class="rc-slider-mark-text rc-slider-mark-text-active"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 66.66666666666666%;"
    >
      €€€
    </span>
    <span
      class="rc-slider-mark-text rc-slider-mark-text-active"
      style="width: 30.000000000000004%; margin-left: -15.000000000000002%; left: 100%;"
    >
      €€€€
    </span>
  </div>
</div>
`;
